ChangeNotify() for local backend using fsnotify

lawmurray opened the pull request:
#### What is the purpose of this change?

This is an initial implementation of `ChangeNotify()` for the local backend, using [fsnotify](https://github.com/fsnotify/fsnotify). I expect that some further discussion and iteration and testing will be needed, but hope this implementation is a useful starting point for that.

#### Was the change discussed in an issue or in the forum before?

Related issues: #249 #4152

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate. **Ran `rclone test changenotify` but further tests may be warranted yet**
- [ ] I have added documentation for the changes if appropriate.
- [ ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

#### Further notes

fsnotify does not support recursive watching of a directory, and it seems that it will not (fsnotify/fsnotify#18) until there is support for all platforms that it targets. There is some work for specific platforms, such as Linux with inotify, although this is not yet released (fsnotify/fsnotify#472).

So at this stage, rather than relying on fsnotify for recursive support, I've implemented that in rclone itself, using `filepath.WalkDir()` (which is what fsnotify/fsnotify#472 is doing anyway), including watching new directories when created. Watching a directory is initiated before listing and recursing into it, which should eliminate race conditions if changes are made while the watchers are established (but further thinking and testing would be worthwhile!).

I've tested this on Linux only. It will need testing on other platforms.

An interactive test routine is to run `rclone test changenotify . -vv` to watch the current working directory, then perform some file operations and see what is logged. The debug logging is verbose for this purpose (it reports all events; this could be reduced eventually). Something like `mkdir -p a/b/c` makes for an interesting test (turns out that an event is produced only for `a`, so it's necessary to then recursively watch `a` to keep the watch list up to date). Perhaps some fuzzing tests with random timings would be useful for race conditions, which may be platform specific.

I hope this is useful, and very happy for others to help with this as well, of course.

lawmurray commented:
Looking into these failed tests.

lawmurray commented on fstest/fstests/fstests.go:
```diff
@@ -876,18 +877,18 @@ func Run(t *testing.T, opt *Opt) {
 			require.NoError(t, err)
 
 			pollInterval := make(chan time.Duration)
-			dirChanges := map[string]struct{}{}
```
This change fixes some tests for the union backend, as multiple goroutines now write to these maps concurrently.

lawmurray commented:
While not yet ready for a merge, this should be in a good state for feedback and discussion now. I've expanded the comments to explain some of the caveats of implementing `ChangeNotify()` with fsnotify. Assuming those caveats are acceptable, I may need a little help on the failing tests (just focused on the [Linux tests](https://github.com/rclone/rclone/actions/runs/12445623349/job/34747135355?pr=8258) for now):

1. `FAIL: TestMount`: does this assume that `ChangeNotify()` is not implemented? The `changeNotify()` at `rclone/vfs/dir.go:285` looks like it invalidates the cache so that it may be correct for the extra directory to appear now.
2. `FAIL: TestRcPollInterval`: if I understand correctly this test looks to check that the polling interval is respected, but in doing so assumes that the implementation uses polling? This one does not.

Still trying to understand these tests some more.

lawmurray commented:
I now have [all tests passing](https://github.com/lawmurray/rclone/actions/runs/12870466766), just on a different branch to avoid creating more noise here. There are still known issues that need discussion if this is to be successful.

I did make quite a mess of this pull request sorry! I'll get it tidied up, or perhaps just close and open a new one, but hope to have this ready for review soon.

ncw commented:
> I now have [all tests passing](https://github.com/lawmurray/rclone/actions/runs/12870466766), just on a different branch to avoid creating more noise here. There are still known issues that need discussion if this is to be successful.

Great - well done :-)

> I did make quite a mess of this pull request sorry! I'll get it tidied up, or perhaps just close and open a new one, but hope to have this ready for review soon.

What I would do is backup the branch then squash all the commits down to one to clean the history, rebase on master then force push. That will give the best experience to review without opening a new PR.

Can you put a list of open issue that need discussing at that point then we can work out what to do with them.

Thank you

lawmurray commented on backend/local/changenotify_other.go:
```diff
@@ -0,0 +1,275 @@
+//go:build !windows
```
While there is some commonality between the implementation for Windows and other platforms, I've separated them out with conditional compilation to keep the logic clearer on reading.

lawmurray commented on fstest/fstests/fstests.go:
The union backend tests fail due to concurrent writes to maps. I've added a mutex to fix this. An alternative is `sync.Map` but it makes error messages harder to parse when expected vs actual results differ.

lawmurray commented on vfs/rc_test.go:
```diff
@@ -97,7 +98,15 @@ func TestRcPollInterval(t *testing.T) {
 	}
 	out, err := call.Fn(context.Background(), nil)
 	require.NoError(t, err)
-	assert.Equal(t, rc.Params{}, out)
```
My understanding of this test is that it's just checking the reported polling interval. That changes for a backend with `ChangeNotify` enabled, so I have updated the expected output to match.

lawmurray commented:
> What I would do is backup the branch then squash all the commits down to one to clean the history, rebase on master then force push. That will give the best experience to review without opening a new PR.
> 
> Can you put a list of open issue that need discussing at that point then we can work out what to do with them.

Thanks @ncw. I've done all that and I think it's now ready for review. I've deleted posts of mine that are no longer relevant and updated my initial post with a detailed description of the known issues that need discussion. Do let me know if you have questions on any of that.

darthShadow commented:
Should we consider adding a dedupe buffer for write events? Otherwise writing of large files will create a lot of events, even with a larger notify interval.

Ideally we would have this PR (https://github.com/fsnotify/fsnotify/pull/651) merged already, so it's handled automatically, atleast for Linux, but since it's not, we will need to implement the dedupe ourselves similar to the logic here: https://github.com/fsnotify/fsnotify/blob/main/cmd/fsnotify/closewrite.go

ncw reviewed (commented):
I put a few comments on the code inline.

I think probably my major concern is performance. We are putting recursive listeners on all the directories in the Linux code. That is going to use a lot of resources.

When I try it on my home directory it uses over 1GB of memory and the watchers start failing with

```
2025/02/26 18:04:53 ERROR : Local file system at /home/ncw: Failed to start watching go/pkg/mod/golang.org/x/tools@v0.23.0/go/ssa/testdata/src/encoding/xml, already removed? no space left on device
```

For an industrial strength tool like rclone, I think it should be able to monitor any size of directory.

Currently there are only two users of ChangeNotify

- cmd/test/changenotify/changenotify.go
- vfs/vfs.go

The VFS use being the most important - it is used to invalidate directories in the directory cache.

So I would certainly consider changing the interface if that makes it easier to implement.

If we do want to use it for bisync or continuous syncing we should have a think about what interface we want there too.

lawmurray commented:
@darthShadow

> Should we consider adding a dedupe buffer for write events? Otherwise writing of large files will create a lot of events, even with a larger notify interval.

Sure. If I understand the logic in the link, it's basically, "don't send a write event if another write event happens in the next 100ms". The code looks to launch a goroutine for every path, I'll try and explore working it into the polling implementation instead, so many goroutines looks a little heavy for the purpose to me.

lawmurray commented on backend/local/changenotify_windows.go:
```diff
@@ -0,0 +1,222 @@
+// Setting fsnotify.enableRecurse to true enables recursive handling: paths
+// that end with with \... or /... as watched recursively.
+
+//go:linkname enableRecurse github.com/fsnotify/fsnotify.enableRecurse
```
Was not aware that it has been deprecated actually, but definitely aware of its naughtiness :laughing: 

It's not ideal, obviously, and meant to be a temporary hack to keep moving. Let me see where upstream is at on potentially providing a public API for this. Looks like they have at least one recent question on the topic (fsnotify/fsnotify#673).

lawmurray commented:
@ncw

> When I try it on my home directory it uses over 1GB of memory and the watchers start failing with
> 
> ```
> 2025/02/26 18:04:53 ERROR : Local file system at /home/ncw: Failed to start watching go/pkg/mod/golang.org/x/tools@v0.23.0/go/ssa/testdata/src/encoding/xml, already removed? no space left on device
> ```
I'll have to profile, but I think the memory use here is likely for the `known` map rather than for the watchers. The `known` map is keyed by the full path of each file and directory, whereas a more memory-efficient tree-like data structure that mimics the file system would reduce memory use a lot. Even better though, the only reason `known` exists at all is bookkeeping for the second argument to `ChangeNotify()`, so an interface change could get rid of it completely and I think there won't be this memory issue at all (which I agree is bad btw, 1 GB is way too much---just didn't want to optimize this data structure until having the conversation on interface).

The error I'll need to check out... `no space left on device` seems odd, unless that relates to the max number of watchers with inotify. This is, unfortunately, always an issue on Linux, I've even seen it with the official Dropbox client. You can (writing here for testing purposes, but should definitely be included in documentation down the line):

1. Check how many directories that will need watchers (one watcher is required per directory, files don't matter):
    ```
    find . -type d | wc -l
    ```
2. Work out maximum number of watchers allowed at present:
    ```
    sudo sysctl -a | grep fs.inotify.max
    ````
    giving by default on e.g. Ubuntu 24.04
    ```
    fs.inotify.max_user_instances = 128  
    fs.inotify.max_user_watches = 65536
    ```
    I believe the former is basically the number of processes that can create watchers, the latter the number of files/directories watched.

    (Incidentally, other distributions, like Arch, are more generous, configured with 1024/524288 by default).

3. Increase the allowed number temporarily:
    ```
    sudo sysctl -n fs.inotify.max_user_instances=1024
    sudo sysctl -n fs.inotify.max_user_watches=524288
    ```
    or write the same to `/etc/sysctl.conf` to make it permanent.

> So I would certainly consider changing the interface if that makes it easier to implement.
>
> If we do want to use it for bisync or continuous syncing we should have a think about what interface we want there too.

I'll follow up with some proposed next steps.

lawmurray commented:
By way of next step, what I'm thinking to do is:

1. Add a new value to the enum `EntryType`, say `EntryUncertain`, for use when a backend cannot reliably or efficiently determine the entry type. The idea is to not require code changes for backends that have an existing implementation that can reliably identify the entry type.
2. Update tests to allow the use of `EntryUncertain`. Some of them collect notifications separately for `EntryObject` and `EntryDirectory`. Ideally this is preserved and the tests will still check that the entry type is correct, if used, but will also allow the use of `EntryUncertain`.
3. Update the local backend to use `EntryUncertain`, see where we're at in terms of memory use, etc.

Sound okay? I understand this won't address everything yet, but it seems like a good next step to me.
